Automattic/jetpack
# Boost: handle post edits and comments and delete cache files

The Cache module has to delete cache files when posts are edited, comments are made or other actions are taken that modify the website.
This PR will take care of those actions and delete the right files so cache files don't grow stale.

Fixes [#35055](https://github.com/Automattic/jetpack/issues/35055)

## Proposed changes:
* Lots of error_log statements still in the code to aid testing. I'll have to remove them, and get an approval again.
* Hook onto various actions so the cache can be cleared when posts are edited, comments made or the theme is switched. We'll probably need more.
* In is_cacheable() check if the user is viewing a preview, or using the customizer. Not applicable right now because we don't cache for logged in users. Maybe in future?
* In normalize_request_uri(), parse_url may return null or false, so we need to handle those.
* Added the function is_visible_post_type( $post ) that returns true if the post type of that post is visible.
* Add the function delete_front_page_cache( $post ) because the front page should be handled separately. It can be in a number of formats (static page, or page of blog posts..)
* Added functions to handle comment editing and publishing.
* The post status can be either published or private which are visible on the web page, or a number of other statuses which aren't.  The delete_on_post_transition() function should take care of that and will delete the related tags and categories, and the homepage.
* delete_cache_for_post() was added to accept a $post, get a permalink, even if the post was trashed, and then delete the cache files for that post.
* delete_cache_for_post_terms() will delete the tag and category archives for the given post.
* delete_cache_for_url( $url ) deletes the cache files for that url, normalizing and then passing on the URI to the storage class.
* delete_cache() will delete the entire cache for the site.
* In the Utils library, the delete_directory() function gets a new parameter, $recurse. If it's 1 then it will delete any directories found by recursing down through them.
* delete_single_directory() simply calls delete_directory() with $recurse set to false.
* In the storage class, invalidate( $path ) will delete the path, recursively.
* invalidate_single_visitor() will delete the cache files for the given path, and user with GET and COOKIEs.
* invalidate_home_page() deletes the html files found in the root of the cache directory, and any "page" archives.

### Other information:


## Jetpack product discussion
pc9hqz-2rA-p2

## Does this pull request change what data or activity we track or use?
no

## Testing instructions:
* Create some tags.
* Create some test posts and add them to the tags.
* Populate the cache by visiting these posts, and the homepage too. You must use a private, NOT LOGGED IN, browser window.
* Tail the logs with `jetpack docker tail`.
* Edit the cache files and watch the tailed log. It should show it deleting that post, and the home page. Please verify the cache files were deleted from the directory.
* Change a post from published to draft. Change it from either of those to trashed. Restore it from trash. No cache files should be edited if the post isn't published either before or after.
* Leave a comment on a post. If moderated, you'll see a log entry saying the cache for that particular user was deleted.
* Try again.
* In wp-admin->comments, moderate the comments. Approve and reverse that. Edit the comment. Trash it. Examine the logs.
* Switch theme and note it deletes the entire cache.
* Link to another post on your site while editing a post. The moderated pingback will appear in the comments page. Approve it and note the cached pages being deleted.
